core/src/main/java/io/opentelemetry/android/export/DelegatingExporter.kt

```diff
@@ -0,0 +1,161 @@
+     *   has been exported. If all of the data was dropped then a failure is returned.
+     */
+    fun export(data: Collection<T>): CompletableResultCode =
+        withDelegate(
+            ifSet = { doExport(this, data) },
+            ifNotSet = {
+                val amountToTake = maxBufferedData - buffer.size
+                buffer.addAll(data.take(amountToTake))
+                // If all the data was dropped we return an exception
+                if (amountToTake == 0 && data.isNotEmpty()) {
+                    CompletableResultCode.ofExceptionalFailure(BufferOverflowException())
+                } else {
+                    pendingExport
+                        ?: CompletableResultCode().also { pendingExport = it }
+                }
+            },
+        )
+
+    /**
+     * Flushes the exporter using the [doFlush] lambda. If the delegate is not yet set a flush will
+     * be scheduled and executed when the delegate is set.
+     *
+     * @return the result. If the delegate is set then the result from it will be returned,
+     *   otherwise a result is returned which will complete when the delegate is set and has been
+     *   flushed.
+     */
+    fun flush(): CompletableResultCode =
+        withDelegate(
+            ifSet = doFlush,
+            ifNotSet = { pendingFlush ?: CompletableResultCode().also { pendingFlush = it } },
+        )
+
+    /**
+     * Shuts down the exporter using the [doShutdown]. If the delegate is not yet set a shut down
+     * will be scheduled and executed when the delegate is set.
+     *
+     * @return the result. If the delegate is set then the result from it will be returned,
+     *   otherwise a result is returned which will complete when the delegate is set and has been
+     *   shut down.
+     */
+    fun shutdown(): CompletableResultCode =
+        withDelegate(
+            ifSet = doShutdown,
+            ifNotSet = { pendingShutdown ?: CompletableResultCode().also { pendingShutdown = it } },
+        )
+
+    private fun clearBuffer() {
+        buffer.clear()
+        buffer.trimToSize()
+    }
+
+    private inline fun <R> withDelegate(
+        ifSet: D.() -> R,
+        ifNotSet: () -> R,
+    ): R {
+        val delegate =
+            synchronized(lock) {
+                delegate ?: return ifNotSet()
+            }
+        // We interact with the delegate outside of the synchronized block to avoid any potential
+        // deadlocks due to reentrant calls
+        return delegate.ifSet()
+    }
+
+    private fun CompletableResultCode.setTo(other: CompletableResultCode) {
+        other.whenComplete {
+            if (other.isSuccess) {
+                succeed()
+            } else {
+                val throwable = other.failureThrowable
+                if (throwable == null) {
+                    fail()
+                } else {
+                    failExceptionally(throwable)
+                }
```

Comment:
I'm pretty sure this can be replaced with
```suggestion
                failExceptionally(throwable)
```
But the docs for `failExceptionally` does not mention nullability so I opted to be safe

Reply:
Yeah that's fair. [The implementation](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/common/src/main/java/io/opentelemetry/sdk/common/CompletableResultCode.java#L129) looks like it would be fine as well, but I think without a `@Nullable` on the param or any javadoc about it, better to play it safe.

Reply:
I also opened https://github.com/open-telemetry/opentelemetry-java/pull/6963 for our future selves. :)